plan/plan_test.go

```diff
@@ -17,13 +17,160 @@ limitations under the License.
 package plan
 
 import (
-	"fmt"
 	"testing"
 
+	"fmt"
 	"github.com/kubernetes-incubator/external-dns/endpoint"
 	"github.com/kubernetes-incubator/external-dns/internal/testutils"
+	"github.com/stretchr/testify/suite"
 )
 
+type PlanTestSuite struct {
+	suite.Suite
+	fooV1Cname *endpoint.Endpoint
+	fooV2Cname *endpoint.Endpoint
+	fooA       *endpoint.Endpoint
+	bar127A    *endpoint.Endpoint
+	bar192A    *endpoint.Endpoint
+}
+
+func (suite *PlanTestSuite) SetupTest() {
+	suite.fooV1Cname = &endpoint.Endpoint{
+		DNSName:    "foo",
+		Target:     "v1",
+		RecordType: "CNAME",
+	}
+	suite.fooV2Cname = &endpoint.Endpoint{
+		DNSName:    "foo",
+		Target:     "v2",
+		RecordType: "CNAME",
+	}
+	suite.bar127A = &endpoint.Endpoint{
+		DNSName:    "bar",
+		Target:     "127.0.0.1",
+		RecordType: "A",
+	}
+	suite.bar192A = &endpoint.Endpoint{
+		DNSName:    "bar",
+		Target:     "192.168.0.1",
+		RecordType: "A",
+	}
+	suite.fooA = &endpoint.Endpoint{
+		DNSName:    "foo",
+		Target:     "5.5.5.5",
+		RecordType: "A",
+	}
+}
+
+func (suite *PlanTestSuite) TestSyncFirstRound() {
+	current := []*endpoint.Endpoint{}
+	desired := []*endpoint.Endpoint{suite.fooV1Cname, suite.fooV2Cname, suite.bar127A}
+	expectedCreate := []*endpoint.Endpoint{suite.fooV1Cname, suite.fooV2Cname, suite.bar127A}
+	expectedUpdateOld := []*endpoint.Endpoint{}
+	expectedUpdateNew := []*endpoint.Endpoint{}
+	expectedDelete := []*endpoint.Endpoint{}
+
+	p := &Plan{
+		Policies: []Policy{&SyncPolicy{}},
+		Current:  current,
+		Desired:  desired,
+	}
+
+	changes := p.Calculate().Changes
+	validateEntries(suite.T(), changes.Create, expectedCreate)
+	validateEntries(suite.T(), changes.UpdateNew, expectedUpdateNew)
+	validateEntries(suite.T(), changes.UpdateOld, expectedUpdateOld)
+	validateEntries(suite.T(), changes.Delete, expectedDelete)
+}
+
+func (suite *PlanTestSuite) TestSyncSecondRound() {
+	current := []*endpoint.Endpoint{suite.fooV1Cname}
+	desired := []*endpoint.Endpoint{suite.fooV2Cname, suite.fooV1Cname, suite.bar127A}
+	expectedCreate := []*endpoint.Endpoint{suite.fooV2Cname, suite.bar127A}
+	expectedUpdateOld := []*endpoint.Endpoint{}
+	expectedUpdateNew := []*endpoint.Endpoint{}
+	expectedDelete := []*endpoint.Endpoint{}
+
+	p := &Plan{
+		Policies: []Policy{&SyncPolicy{}},
+		Current:  current,
+		Desired:  desired,
+	}
+
+	changes := p.Calculate().Changes
+	validateEntries(suite.T(), changes.Create, expectedCreate)
+	validateEntries(suite.T(), changes.UpdateNew, expectedUpdateNew)
+	validateEntries(suite.T(), changes.UpdateOld, expectedUpdateOld)
+	validateEntries(suite.T(), changes.Delete, expectedDelete)
+}
+
+func (suite *PlanTestSuite) TestIdempotency() {
+	current := []*endpoint.Endpoint{suite.fooV1Cname, suite.fooV2Cname}
```

Comment:
the reason why I started this was because service source already creates several records with the same DNSName and type and this need to be handled somehow